pkg/sfu/connectionquality/connectionstats.go

```diff
@@ -1,57 +1,205 @@
+	return &ConnectionStats{
+		params: params,
+		score:  4.0,
+		done:   make(chan struct{}),
+	}
+}
+
+func (cs *ConnectionStats) Start() {
+	go cs.updateStats()
+}
+
+func (cs *ConnectionStats) Close() {
+	if !cs.isClosed.TrySet(true) {
+		return
+	}
+
+	close(cs.done)
 }
 
-func NewConnectionStats() *ConnectionStats {
-	return &ConnectionStats{Prev: &ConnectionStat{}, Score: 4.0}
+func (cs *ConnectionStats) OnStatsUpdate(fn func(cs *ConnectionStats, stat *livekit.AnalyticsStat)) {
+	cs.onStatsUpdate = fn
 }
 
-func getTotalPackets(curSN, prevSN uint32) uint32 {
-	delta := curSN - prevSN
-	// lower 16 bits is the counter
-	counter := uint16(delta)
-	// upper 16 bits contains the cycles/wrap
-	cycles := uint16(delta >> 16)
-	increment := (uint32(cycles) * (1 << 16)) + uint32(counter)
+func (cs *ConnectionStats) RTCPFeedback(packets []rtcp.Packet, expectedSSRC uint32) {
+	if cs.isClosed.Get() {
+		return
+	}
+
+	cs.lock.Lock()
+	defer cs.lock.Unlock()
+
+	for _, p := range packets {
+		switch pkt := p.(type) {
+		case *rtcp.ReceiverReport:
+			for _, r := range pkt.Reports {
+				if expectedSSRC != 0 && r.SSRC != expectedSSRC {
+					continue
+				}
+
+				if r.Delay > cs.maxDelay {
+					cs.maxDelay = r.Delay
+				}
+
+				if r.Jitter > cs.maxJitter {
+					cs.maxJitter = r.Jitter
+				}
+
+				if !cs.baseSeqNumInitialized {
+					cs.baseSeqNumInitialized = true
+					cs.baseSeqNum = r.LastSequenceNumber
+					cs.highestSeqNum = r.LastSequenceNumber
+					cs.totalPacketsLost = r.TotalLost
+				}
+
+				if r.LastSequenceNumber > cs.highestSeqNum {
+					cs.highestSeqNum = r.LastSequenceNumber
+					cs.totalPacketsLost = r.TotalLost
+				}
+			}
+
+		case *rtcp.TransportLayerNack:
+			nackCount := 0
+			for _, pair := range pkt.Nacks {
+				nackCount += len(pair.PacketList())
+			}
+			cs.nackCount += int32(nackCount)
+
+		case *rtcp.PictureLossIndication:
+			cs.pliCount += 1
 
-	return increment
+		case *rtcp.FullIntraRequest:
+			cs.firCount += 1
+		}
+	}
 }
 
-func (cs *ConnectionStats) UpdateStats(totalBytes uint64) ConnectionStat {
-	// update feedback stats
-	previous := cs.Prev
+func (cs *ConnectionStats) GetScore() float64 {
+	cs.lock.RLock()
+	defer cs.lock.RUnlock()
+
+	return cs.score
+}
+
+func (cs *ConnectionStats) updateAndGetPercentageLoss() float64 {
+	if cs.params.GetTotalBytes != nil {
+		cs.totalBytes = cs.params.GetTotalBytes()
+	}
+
+	if !cs.snapshot.Initialized {
+		cs.snapshot.Initialized = true
+		cs.snapshot.HighestSeqNum = cs.highestSeqNum
+		cs.snapshot.TotalPacketsLost = cs.totalPacketsLost
+	}
+
+	packetsLostInInterval := cs.totalPacketsLost - cs.snapshot.TotalPacketsLost
+	expectedPacketsInInterval := cs.highestSeqNum - cs.snapshot.HighestSeqNum
+	percentageLoss := float64(0.0)
+	if expectedPacketsInInterval > 0 {
+		percentageLoss = (float64(packetsLostInInterval) / float64(expectedPacketsInInterval)) * 100
+	}
+
+	cs.snapshot.HighestSeqNum = cs.highestSeqNum
```

Comment:
@boks1971 I had noticed that some of these rtcp reports were being retransmitted. We should ideally only update/use only if cs.highestSeqNum > snaphsot.HighestSeqNum